nemo/core/optim/lr_scheduler.py

```diff
@@ -0,0 +1,412 @@
+        self.min_lr = min_lr
+        self.power = power
+        self.cycle = cycle
+
+        super().__init__(optimizer=optimizer, max_steps=max_steps, last_epoch=last_epoch, **kwargs)
+
+    def _get_lr(self, step):
+        new_lrs = [
+            _poly_decay(
+                initial_lr,
+                step=step - self.hold_steps,
+                decay_steps=self.max_steps - max(self.warmup_steps, self.hold_steps),
+                power=self.power,
+                min_lr=self.min_lr,
+                cycle=self.cycle,
+            )
+            for initial_lr in self.base_lrs
+        ]
+        return new_lrs
+
+
+def prepare_lr_scheduler(
+    optimizer: optim.Optimizer,
+    scheduler_config: Dict[str, Any],
+    train_dataloader: Optional[dataloader.DataLoader] = None,
+) -> Optional[Dict[str, Any]]:
+    """
+    Constructs an LR Scheduler (optionally) for a given optimizer, based on a config with the following schema
+
+    {
+    ...,
+    "scheduler": <a class that inherits torch.optim.lr_scheduler._LRScheduler>,
+    "scheduler_args": {
+        "max_steps": int, <OR> "iters_per_batch": int,
+        "monitor": <metric to monitor; say "loss" or "val_loss">,
+        <any kwarg to pass onto the optimizer>
+      }
+    }
+
+    Args:
+        optimizer: An instantiated Optimizer.
+        scheduler_config: A dictionary which follows the above schema.
+        train_dataloader: Optional requirement, must be passed if "iters_per_batch" is defined
+            instead of "max_steps". Used to compute effective "max_steps".
+
+    Returns:
+        A dictionary containing the LR Scheduler implementation if the config was successfully parsed
+        along with other parameters required by Pytorch Lightning, otherwise None.
+    """
+    if 'scheduler' in scheduler_config:
+        if 'scheduler_args' in scheduler_config:
+            scheduler_args = scheduler_config['scheduler_args']
+        else:
+            raise ValueError("If `scheduler` is provided, `scheduler_args` must be provided.")
+
+    else:
+        logging.info('Scheduler not initialized as no `scheduler` argument supplied to setup_optimizer()')
+        return None
+
+    # Get the scheduler class from the config
+    scheduler = scheduler_config['scheduler']
+
+    # Extract value to monitor in losses, if provided.
+    if 'monitor' in scheduler_args:
+        monitor = scheduler_args.pop('monitor')
+    else:
+        # default to train loss
+        monitor = 'loss'
+
+    # Compute effective max_steps if iters_per_batch is provided
+    if 'iters_per_batch' in scheduler_args:
+        if train_dataloader is None:
+            raise ValueError(
+                'As `iters_per_batch` is provided, it is required to pass the train dataloader in order '
+                'to compute effective maximum number of steps'
+            )
+
+        iters_per_batch = scheduler_args.pop('iters_per_batch')
+        num_samples = len(train_dataloader.dataset)
```

Comment:
+@redoctopus 
Is this going to crash Iterable datasets? Since they implement iter() and not getitem()/len()

Reply:
I checked before, https://github.com/NVIDIA/NeMo/blob/master/nemo/collections/asr/data_layer.py#L583, but it seems TarredAudioTextLayer does implement Len(). @redoctopus can you confirm this works? And after discussing with @yzhang123, I found that most IterableDataset require you to pass max_steps anyway instead of allowing the computation of effective max steps (since they may not implement __len__)